src/backtracking/all_combination_of_size_k.rs

```diff
@@ -1,62 +1,121 @@
-/*
-    In this problem, we want to determine all possible combinations of k
-    numbers out of 1 ... n. We use backtracking to solve this problem.
-    Time complexity: O(C(n,k)) which is O(n choose k) = O((n!/(k! * (n - k)!)))
-
-    generate_all_combinations(n=4, k=2) => [[1, 2], [1, 3], [1, 4], [2, 3], [2, 4], [3, 4]]
-*/
-pub fn generate_all_combinations(n: i32, k: i32) -> Vec<Vec<i32>> {
-    let mut result = vec![];
-    create_all_state(1, n, k, &mut vec![], &mut result);
-
-    result
+//! This module provides a function to generate all possible combinations
+//! of `k` numbers out of `1...n` using a backtracking algorithm.
+
+/// Custom error type for combination generation.
+#[derive(Debug, PartialEq)]
+pub enum CombinationError {
+    KGreaterThanN,
+    InvalidZeroRange,
+}
+
+/// Generates all possible combinations of `k` numbers out of `1...n`.
+///
+/// # Arguments
+///
+/// * `n` - The upper limit of the range (1 to `n`).
+/// * `k` - The number of elements in each combination.
+///
+/// # Returns
+///
+/// A `Result` containing a vector with all possible combinations of `k` numbers out of `1...n`,
+/// or a `CombinationError` if the input is invalid.
+pub fn generate_all_combinations(n: usize, k: usize) -> Result<Vec<Vec<usize>>, CombinationError> {
+    if n == 0 && k > 0 {
+        return Err(CombinationError::InvalidZeroRange);
+    }
+
+    if k > n {
+        return Err(CombinationError::KGreaterThanN);
+    }
+
+    let mut combinations = vec![];
+    backtrack(1, n, k, &mut vec![], &mut combinations);
```

Comment:
```suggestion
    let mut current = vec![0; k];
    backtrack(0, n, k, 0, &mut current, &mut combinations); // Start from 0 for zero-indexing
```
The recursion starts from `start = 0`, so it aligns with zero-indexing.